odoo/tools/lru.py

```diff
@@ -41,88 +29,50 @@ def get(self, obj, val=None):
 
     @synchronized()
     def __getitem__(self, obj):
-        a = self.d[obj].me
-        self[a[0]] = a[1]
-        return a[1]
+        a = self.d[obj]
+        self.d.move_to_end(obj, last=False)
+        return a
 
     @synchronized()
     def __setitem__(self, obj, val):
-        if obj in self.d:
-            del self[obj]
-        nobj = LRUNode(self.last, (obj, val))
-        if self.first is None:
-            self.first = nobj
-        if self.last:
-            self.last.next = nobj
-        self.last = nobj
-        self.d[obj] = nobj
-        if len(self.d) > self.count:
-            if self.first == self.last:
-                self.first = None
-                self.last = None
-                return
-            a = self.first
-            a.next.prev = None
-            self.first = a.next
-            a.next = None
-            del self.d[a.me[0]]
-            del a
+        self.d[obj] = val
+        self.d.move_to_end(obj, last=False)
+        while len(self.d) > self.count:
+            self.d.popitem(last=True)
 
     @synchronized()
     def __delitem__(self, obj):
-        nobj = self.d[obj]
-        if nobj.prev:
-            nobj.prev.next = nobj.next
-        else:
-            self.first = nobj.next
-        if nobj.next:
-            nobj.next.prev = nobj.prev
-        else:
-            self.last = nobj.prev
         del self.d[obj]
 
     @synchronized()
     def __iter__(self):
-        cur = self.first
-        while cur is not None:
-            cur2 = cur.next
-            yield cur.me[1]
-            cur = cur2
+        yield from self.d
 
     @synchronized()
     def __len__(self):
         return len(self.d)
 
-    # FIXME: should this have a P2 and a P3 version or something?
     @synchronized()
     def iteritems(self):
-        cur = self.first
-        while cur is not None:
-            cur2 = cur.next
-            yield cur.me
-            cur = cur2
+        yield from self.d.items()
     items = iteritems
 
     @synchronized()
     def iterkeys(self):
-        return iter(self.d)
+        yield from self.d
 
     @synchronized()
     def itervalues(self):
-        return iter(self.d.values())
+        yield from self.d.values()
```

Comment:
Should fix synchronisation issue: the creation of the `values()` iterator would be synchronised but the iteration itself wouldn't be.